File: code-ninja-course/src/main/java/com/codeninja/solidsolution/BookAllowanceSetter.java
package com.codeninja.solidsolution;

import java.util.HashMap;

import com.codeninja.solidsolution.rules.*;
import com.codeninja.solidsolution.students.Student;

public class BookAllowanceSetter {
	Student student;
	HashMap<PackageType, AllowanceRule> rules;
	
	public BookAllowanceSetter(Student student) {
		this.student = student;
		
		rules = new HashMap<PackageType, AllowanceRule>() {
			{
				put(PackageType.Premium, new PremiumAllowanceRule());
				put(PackageType.Standard, new StandardAllowanceRule());
			}
		};
	}
	
	public void setAllowance(PackageType type) {
		int factor = rules.get(type).getFactor();
		student.setMonthlyEbookAllowance(factor * 10);
	}
	
}

File: code-ninja-course/CSHARP/Code/SOLIDSolution/BookAllowanceSetter.cs
using System.Collections.Generic;

namespace LegacyCodeToTestableCode.SOLIDSolution
{
    public class BookAllowanceSetter
    {
        Student student;
        Dictionary<PackageType, IAllowanceRule> rules;

        public BookAllowanceSetter(Student student)
        {
            this.student = student;

            rules = new Dictionary<PackageType, IAllowanceRule>()
            {
                {PackageType.Premium, new PremiumAllowanceRule() },
                {PackageType.Standard, new StandardAllowanceRule()}
            };
        }


        public void SetAllowance(PackageType type)
        {
            int factor = rules[type].getFactor();
            student.setMonthlyEbookAllowance(factor * 10);
        }
    }
}
